syft/source/ocimodelsource/oci_model_source.go

```diff
@@ -18,95 +21,119 @@ import (
+	Alias        source.Alias
 }
 
 // ociModelSource implements the source.Source interface for OCI model artifacts.
 type ociModelSource struct {
-	id       artifact.ID
-	config   Config
-	resolver *ociModelResolver
-	mutex    *sync.Mutex
+	id         artifact.ID
+	reference  string
+	alias      source.Alias
+	metadata   *OCIModelMetadata
+	tempDir    string
+	layerFiles map[string]layerInfo
+	resolver   file.Resolver
+	mutex      *sync.Mutex
 }
 
-// NewFromArtifact creates a new OCI model source from a fetched model artifact.
-func NewFromArtifact(artifact *ModelArtifact, client *RegistryClient, alias source.Alias) (source.Source, error) {
-	// Build metadata
+// NewFromRegistry creates a new OCI model source by fetching the model artifact from a registry.
+func NewFromRegistry(ctx context.Context, cfg Config) (source.Source, error) {
+	client, err := NewRegistryClient(cfg.RegistryOpts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create registry client: %w", err)
+	}
+
+	artifact, err := validateAndFetchArtifact(ctx, client, cfg.Reference)
+	if err != nil {
+		return nil, err
+	}
+
 	metadata := buildMetadata(artifact)
+	tempDir, layerFiles, err := fetchAndStoreGGUFHeaders(ctx, client, artifact)
+	if err != nil {
+		return nil, err
+	}
 
-	// Fetch GGUF layer headers via range-GET
-	tempFiles := make(map[string]string)
-	ggufLayers := make([]GGUFLayerInfo, 0, len(artifact.GGUFLayers))
+	id := deriveID(cfg.Reference, cfg.Alias, metadata.ManifestDigest)
+	return &ociModelSource{
+		id:         id,
+		reference:  cfg.Reference,
+		alias:      cfg.Alias,
+		metadata:   metadata,
+		tempDir:    tempDir,
+		layerFiles: layerFiles,
+		mutex:      &sync.Mutex{},
+	}, nil
+}
 
-	for idx, layer := range artifact.GGUFLayers {
-		log.WithFields("digest", layer.Digest, "size", layer.Size).Debug("fetching GGUF layer header")
+// validateAndFetchArtifact fetches and validates a model artifact in a single registry call.
+func validateAndFetchArtifact(ctx context.Context, client *RegistryClient, reference string) (*ModelArtifact, error) {
+	artifact, err := client.FetchModelArtifact(ctx, reference)
+	if err != nil {
+		// ErrNotModelArtifact is wrapped, so callers can use errors.Is() to check
+		return nil, err
+	}
 
-		// Fetch header via range-GET
-		headerData, err := client.FetchBlobRange(context.Background(), artifact.Reference, layer.Digest, MaxHeaderBytes)
-		if err != nil {
-			return nil, fmt.Errorf("failed to fetch GGUF layer header: %w", err)
-		}
+	if len(artifact.GGUFLayers) == 0 {
+		return nil, fmt.Errorf("model artifact has no GGUF layers")
+	}
+
+	return artifact, nil
+}
 
-		// Extract virtual path from annotations
-		virtualPath := extractVirtualPath(idx)
+// fetchAndStoreGGUFHeaders fetches GGUF layer headers and stores them in temp files.
+func fetchAndStoreGGUFHeaders(ctx context.Context, client *RegistryClient, artifact *ModelArtifact) (string, map[string]layerInfo, error) {
+	tempDir, err := os.MkdirTemp("", "oci-gguf")
```

Comment:
```suggestion
tempDir, err := os.MkdirTemp("", "syft-oci-gguf")
```
makes it easier to understand un crashes what resources are ours